examples/knative/python/kn-py-vro/handler.py

```diff
@@ -0,0 +1,270 @@
+        if vroauth == "BASIC":
+            r = requests.post(vroUrl,
+                          auth=(vrouser.encode('utf-8'), vropass.encode('utf-8')),
+                          json=body,
+                          verify=not vroinsecure
+                          )
+        if vroauth == "OAUTH":
+            r = requests.post(vroUrl,
+                          headers={"Authorization":f'Bearer {bearer_token}'},
+                          json=body,
+                          verify=not vroinsecure
+                          )
+            if r.status_code == 401:
+                get_token()
+        if r.ok:
+            success = True
+        if retries < 0:
+            return get_response('vRO REST Request failed > Exception: {0}'.format(r.reason), r.status_code)
+        retries -= 1
+
+    app.logger.debug(f'{bgc.OKBLUE}POST Successful...{bgc.ENDC}')
+
+    try:
+        vro_res = json.loads(r.text)
+    except json.decoder.JSONDecodeError:
+        traceback.print_exc(limit=1, file=sys.stderr)  # providing traceback since it helps debug the exact key that failed
+        return get_response(f'Response is not valid JSON\n{r.text}', r.status_code)
+
+    if r.ok:
+        app.logger.debug(f'Successfully executed vRO workflow: {vro_res["name"]}')
+
+        app.logger.debug(f'{bgc.OKBLUE}vRO Response: {bgc.ENDC}')
+        app.logger.debug(json.dumps(vro_res, indent=4))
+        return get_response(f'Workflow Token {vro_res["id"]}: {vro_res["state"]}', r.status_code)
+    else:
+        app.logger.debug(f'{bgc.FAIL}Failed to execute workflow:{bgc.ENDC}')
+        return get_response(vro_res["message"], vro_res["status"])
+
+
+@app.before_first_request
+def before_first_request():
+    """
+    Runs before any request to perform initialisation 
+    """
+    global vroconfig, vrohost, vroport, vroauth, vrouser, vropass, vrotoken, vroinsecure, wfId
+
+    # Get function config values
+    try:
+        vroconfig = json.loads(os.environ['VROCONFIG_SECRET'])
+        vrohost = vroconfig['SERVER']
+        vroport = vroconfig['PORT']
+        vroauth = vroconfig['AUTH_TYPE'].upper()
+        if vroauth not in ('BASIC', 'OAUTH'):
+            raise KeyError("AUTH_TYPE must be BASIC or OAUTH")
+        if vroauth == 'BASIC' or not vroconfig.get('REFRESH_TOKEN'):
+            vrouser = vroconfig['USERNAME']
+            vropass = vroconfig['PASSWORD']
+            vrotoken = None
+        else:
+            vrotoken = vroconfig['REFRESH_TOKEN']
+        vroinsecure = vroconfig['INSECURE_SSL']
+        wfId = vroconfig['WORKFLOW_ID']
+    except OSError as err:
+        return get_response('Could not read vro configuration: {0}'.format(err), 500)
+    except KeyError as err:
+        return get_response('Mandatory configuration key not found: {0}'.format(err), 500)
+    
+    if(vroinsecure):
+        # Surpress SSL warnings
+        urllib3.disable_warnings(urllib3.exceptions.InsecureRequestWarning)
+    
+    get_token()
```

Comment:
question: what if this throws?

Reply:
Yep, exception handling incoming...!

Reply:
This took some work to get it to work properly within the Flask framework, but hopefully is all OK now. Better handling added around the reading of the config from the secret too.